src/miniscript/astelem.rs

```diff
@@ -58,99 +59,158 @@ impl<Pk: MiniscriptKey> Terminal<Pk> {
     }
 }
 
-impl<Pk: MiniscriptKey> Terminal<Pk> {
+//helper function for Translating two child Terminals
+fn translate_bin<Pk, Ctx, FPk, FPkh, Q, FuncError, TermF>(
+    left: &Arc<Miniscript<Pk, Ctx>>,
+    right: &Arc<Miniscript<Pk, Ctx>>,
+    translatefpk: &mut FPk,
+    translatefpkh: &mut FPkh,
+    term: TermF,
+) -> Result<Result<Terminal<Q, Ctx>, Error>, FuncError>
+where
+    FPk: FnMut(&Pk) -> Result<Q, FuncError>,
+    FPkh: FnMut(&Pk::Hash) -> Result<Q::Hash, FuncError>,
+    Q: MiniscriptKey,
+    Pk: MiniscriptKey,
+    Ctx: ScriptContext,
+    TermF: Fn(Arc<Miniscript<Q, Ctx>>, Arc<Miniscript<Q, Ctx>>) -> Terminal<Q, Ctx>,
+{
+    let l = left
+        .translate_pk(&mut *translatefpk, &mut *translatefpkh)?
+        .map(Arc::new);
+    let r = right
+        .translate_pk(&mut *translatefpk, &mut *translatefpkh)?
+        .map(Arc::new);
+    match (l, r) {
+        (Ok(l), Ok(r)) => Ok(Ok(term(l, r))),
+        (Err(e), _) => Ok(Err(e)),
+        (_, Err(e)) => Ok(Err(e)),
+    }
+}
+
+impl<Pk: MiniscriptKey, Ctx: ScriptContext> Terminal<Pk, Ctx> {
     /// Convert an AST element with one public key type to one of another
     /// public key type
-    pub fn translate_pk<FPk, FPkh, Q, Error>(
+    pub fn translate_pk<FPk, FPkh, Q, FuncError>(
         &self,
         translatefpk: &mut FPk,
         translatefpkh: &mut FPkh,
-    ) -> Result<Terminal<Q>, Error>
+    ) -> Result<Result<Terminal<Q, Ctx>, Error>, FuncError>
```

Comment:
Here too we can drop the double-result (and probably leave the old logic in rather than calling out to `translate_bin`?)